### PR TITLE
Move actors outside of the desired footprint when placing a building

### LINKED ISSUE #3682: Smart Infantry Units : they could move to let the player install a building [$30]
What about infantry units that automatically move if the player tries to install a building where they are standing ?

It would just be a short move, just to let the building being placed.

<bountysource-plugin>
---
There is a **[$30 open bounty](https://www.bountysource.com/issues/792109-smart-infantry-units-they-could-move-to-let-the-player-install-a-building-15?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by penev92**
What the title says.
Spiritual successor to #7800.
Closes #3682.

Tested with 3 clients in a 2v2 match (+ 1 AI) and couldn't get anything to break or behave exploit-y.
You **cannot** tell allied units to move. This is a very basic engine design choice to prevent exploits - orders are accepted only if the client that issued them is the one that has the owner of the order's subject. See [here](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/Traits/ValidateOrder.cs#L37).

### CODE REVIEW

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 1]
```diff
@@ -65,7 +67,11 @@ public IEnumerable<Order> Order(World world, CPos xy, MouseInput mi)
 				world.CancelInputMode();
 
 			var ret = InnerOrder(world, xy, mi).ToArray();
-			if (ret.Length > 0)
+
+			// If there was a successful placement order
+			if (ret.Any(o => o.OrderString == "PlaceBuilding"
+				|| o.OrderString == "LineBuild"
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Personal preference (and I'm pretty sure we do this elsewhere in the code): can you move the `||`s to the end of the preceding line?

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 2]
```diff
@@ -210,5 +219,25 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr, World world)
 		}
 
 		public string GetCursor(World world, CPos xy, MouseInput mi) { return "default"; }
+
+		IEnumerable<Order> TryClearBlockers(World world, CPos topLeft)
+		{
+			var allTiles = FootprintUtils.Tiles(world.Map.Rules, building, buildingInfo, topLeft).ToArray();
+			var neightborTiles = Util.ExpandFootprint(allTiles, true).Except(allTiles)
+				.Where(t => world.Map.Contains(t)).ToList();
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: You ought to be able to remove the lambda from this and have `.Where(world.Map.Contains)` for a tiny bit less overhead.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 3]
```diff
@@ -210,5 +219,25 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr, World world)
...
+
+			var blockers = allTiles.SelectMany(t => world.ActorMap.GetUnitsAt(t))
+				.Where(a => a.Owner == producer.Owner && a.HasTrait<Mobile>());
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Is `Mobile` necessary, or can we reduce this to `IMove`?

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 4]
```diff
@@ -210,5 +219,25 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr, World world)
...
+
+			foreach (var blocker in blockers)
+			{
+				var availableCells = neightborTiles.Where(t => blocker.Trait<Mobile>().CanEnterCell(t));
+
+				yield return new Order("Move", blocker, false)
+				{
+					TargetLocation = blocker.ClosestCell(availableCells)
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: One issue I ran into was a cell becoming full while actors were still trying to find a destination so they ended up hanging around still.

Of course you can spam click to eventually solve this.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs (hunk 2 above)
Comment (penev92) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Ah right. That had more stuff that got removed, but the lambda remained.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs (hunk 4 above)
Comment (penev92) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Yeah, when dealing with more than a few actors you sometimes need to issue the build order more than once :/

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs (hunk 3 above)
Comment (penev92) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: I started with `IMove`, but `Mobile` gives me the `.CanEnterCell(t)` that I use below, so I figured this won't affect aircraft anyway, might as well use `Mobile`.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs (hunk 1 above)
Comment (penev92) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: I think it's easier to both see how far the expression extends (which is not always obvious at first glance thanks to our indentation rules) and what the operators are. Just makes for easier readability IMO.

**Comment by Unit158** [comment]
I believe you could allow ally units to move if you use PlaceBuildingOrderGenerator to move them to the exact same place on all the clients, but this would be very hacky. This should still be investigated (moving allied units) in the future, I'll open an issue after this is merged.

**Comment by Phrohdoh** [comment]
You could allow it but we shouldn't.
Giving orders to actors that aren't yours is not a polite thing to do.

**Comment by penev92** [comment]
I **could** change `ValidateOrder` to allow orders from allies, but that opens the gates to tons of potential exploits and regressions.

**Comment by Unit158** [comment]
From a gameplay PoV, not having allied soldiers move makes this feature seem half baked, that's all I'm saying.

**Comment by Unit158** [comment]
Another possibility would be to add exceptions for certain orders somehow (like a firewall), but that would be dangerous (and potentially slower) too.

**Comment by Unit158** [comment]
@chrisforbes any opinion on the discussion above?

**Comment by Mailaender** [comment]
> From a gameplay PoV, not having allied soldiers move makes this feature seem half baked, that's all I'm saying.

If you can move your team mates soldiers by continuously placing structures on them, that sounds like it will only get exploited. In most cases it is your own units that just came out of the production buildings to clunk up your base, so I think this is fine as is.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 5]
```diff
@@ -210,5 +219,25 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr, World world)
...
 		public string GetCursor(World world, CPos xy, MouseInput mi) { return "default"; }
+
+		IEnumerable<Order> TryClearBlockers(World world, CPos topLeft)
+		{
+			var allTiles = FootprintUtils.Tiles(world.Map.Rules, building, buildingInfo, topLeft).ToArray();
+			var neightborTiles = Util.ExpandFootprint(allTiles, true).Except(allTiles)
+				.Where(world.Map.Contains).ToList();
+
+			var blockers = allTiles.SelectMany(world.ActorMap.GetUnitsAt)
+				.Where(a => a.Owner == producer.Owner && a.HasTrait<Mobile>());
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Can you add a `IsMoving` check here to not stop an army marching away already or harvesters stopping their business?

**Comment by Mailaender** [comment]
Works as promised and doesn't desync in multiplayer. :white_check_mark:

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs (hunk 5 above)
Comment (penev92) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Sure. I thought I had that.

Comment (pchote) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: `IsIdle` would be more appropriate than `IsMoving`.

**Comment by penev92** [comment]
Updated to not bother actors that aren't idle.

**Comment by abcdefg30** [comment]
Needs a rebase.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 6]
```diff
@@ -210,5 +220,25 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr, World world)
 		}
 
 		public string GetCursor(World world, CPos xy, MouseInput mi) { return "default"; }
+
+		IEnumerable<Order> TryClearBlockers(World world, CPos topLeft)
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Could this be renamed to reflect its return type, please?
The `Try` prefix is typically reserved for methods such as `TryParse*` and is a `void`.

Perhaps `ClearBlockersOrders`.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 7]
```diff
@@ -210,5 +220,25 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr, World world)
...
+			var allTiles = FootprintUtils.Tiles(world.Map.Rules, building, buildingInfo, topLeft).ToArray();
+			var neightborTiles = Util.ExpandFootprint(allTiles, true).Except(allTiles)
+				.Where(world.Map.Contains).ToList();
+
+			var blockers = allTiles.SelectMany(world.ActorMap.GetUnitsAt)
+				.Where(a => a.Owner == producer.Owner && a.IsIdle && a.HasTrait<Mobile>());
+
+			foreach (var blocker in blockers)
+			{
+				var availableCells = neightborTiles.Where(t => blocker.Trait<Mobile>().CanEnterCell(t));
```
Comment (Phrohdoh) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Can you check if this is empty, and `continue` if so?
Not doing so is causing auditory feedback that is actually invalid.

>Moving out!
When there is in fact nowhere to go.

**Comment by Phrohdoh** [comment]
The comment regarding an empty `availableCells` is the change I am requesting.
I'd like to see the other in too, but we have a way of ignoring the comments we don't agree with (or like to label as bikeshedding).

After that, this has my :+1: assuming nothing else is changed.
I haven't been able to test this over a network game, and I think that should be done before merging.

**Comment by penev92** [comment]
Rebased and addressed some more comments.

File: OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs [hunk 8]
```diff
@@ -210,5 +220,28 @@ public IEnumerable<IRenderable> RenderAfterWorld(WorldRenderer wr, World world)
...
+		{
+			var allTiles = FootprintUtils.Tiles(world.Map.Rules, building, buildingInfo, topLeft).ToArray();
+			var neightborTiles = Util.ExpandFootprint(allTiles, true).Except(allTiles)
+				.Where(world.Map.Contains).ToList();
+
+			var blockers = allTiles.SelectMany(world.ActorMap.GetUnitsAt)
+				.Where(a => a.Owner == producer.Owner && a.IsIdle)
+				.Select(a => new TraitPair<Mobile> { Actor = a, Trait = a.TraitOrDefault<Mobile>() });
+
+			foreach (var blocker in blockers.Where(x => x.Trait != null))
```
Comment (pchote) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: Recent perf Prs suggest that there is a non-negligible improvement from using an if () continue; in the loop instead.

Comment (RoosterDragon) [inline OpenRA.Mods.Common/Orders/PlaceBuildingOrderGenerator.cs]: It's a tiny saving that's only worth thinking about in code that gets called many times per second. Placing a building doesn't begin to come close - I wouldn't bother.

**Comment by obrakmann** [comment]
Works as expected in SP, MP and in replays. Code looks sensible as well. :+1: 

We should probably mention somewhere that this will only work for idle actors.

**Comment by obrakmann** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/2326c2b26cd3aff1c4be4a9ddd50380f7f84001f)

**Pull request merged**